File: mobile/shared/src/commonMain/kotlin/uk/foodclassifier/text/ForeignColumnFilter.kt
package uk.foodclassifier.text

/**
 * Removes adjacent-language column markers (e.g. NL-INGREDIENTS) without
 * stripping valid ingredient words from the English column.
 */
internal object ForeignColumnFilter {
    private val foreignColumnHeader =
        Regex("""(?i)\b(?!EN)[A-Z]{2}-ING(?:REDIENTS?|REL)\s*:?\s*""")
    private val hyphenatedForeignFragment = Regex("""\b[a-z]+(?:-[a-z]+)+\b""")

    fun strip(text: String): String {
        return text
            .replace(foreignColumnHeader, " ")
            .replace(hyphenatedForeignFragment, " ")
            .replace(Regex("""\s+ne\s+"""), " ")
            .replace(Regex("""\s+"""), " ")
            .trim()
    }
}

File: flutter/lib/text/foreign_column_filter.dart
import 'regex_utils.dart';

class ForeignColumnFilter {
  ForeignColumnFilter._();

  static final _hyphenatedForeignFragment =
      RegExp(r'\b[a-z]+(?:-[a-z]+)+\b');

  static String strip(String text) {
    var result = text.replaceAllMapped(
      ic(r'\b([A-Z]{2})-ING(?:REDIENTS?|REL)\s*:?\s*'),
      (match) {
        if (match.group(1)!.toUpperCase() == 'EN') {
          return match.group(0)!;
        }
        return ' ';
      },
    );
    return result
        .replaceAll(_hyphenatedForeignFragment, ' ')
        .replaceAll(RegExp(r'\s+ne\s+'), ' ')
        .replaceAll(RegExp(r'[^\S\n]+'), ' ')
        .trim();
  }
}
